src/uptane/tuf.cc

```diff
@@ -0,0 +1,230 @@
+      continue;
+    }
+    // Threshold
+    int requiredThreshold = (*it)["threshold"].asInt();
+    if (requiredThreshold < kMinSignatures) {
+      LOGGER_LOG(LVL_debug, "Failing with threshold for role " << role << " too small: " << requiredThreshold << " < "
+                                                               << kMinSignatures);
+      throw IllegalThreshold(repository, "The role " + StringFromRole(role) + " had an illegal signature threshold.");
+    }
+    if (kMaxSignatures < requiredThreshold) {
+      LOGGER_LOG(LVL_debug, "Failing with threshold for role " << role << " too large: " << kMaxSignatures << " < "
+                                                               << requiredThreshold);
+      throw IllegalThreshold(repository, "root.json contains a role that requires too many signatures");
+    }
+    thresholds_for_role_[role] = requiredThreshold;
+
+    // KeyIds
+    Json::Value keyids = (*it)["keyids"];
+    for (Json::ValueIterator itk = keyids.begin(); itk != keyids.end(); ++itk) {
+      keys_for_role_.insert(std::make_pair(role, (*itk).asString()));
+    }
+  }
+}
+
+Json::Value Root::UnpackSignedObject(TimeStamp now, std::string repository, Role role,
+                                     const Json::Value &signed_object) {
+  if (policy_ == kAcceptAll) {
+    return signed_object["signed"];
+  } else if (policy_ == kRejectAll) {
+    throw SecurityException(repository, "Root policy is kRejectAll");
+  }
+  assert(policy_ == kCheck);
+
+  std::string canonical = Json::FastWriter().write(signed_object["signed"]);
+  Json::Value signatures = signed_object["signatures"];
+  int valid_signatures = 0;
+
+  for (Json::ValueIterator sig = signatures.begin(); sig != signatures.end(); ++sig) {
+    std::string method((*sig)["method"].asString());
+    std::transform(method.begin(), method.end(), method.begin(), ::tolower);
+
+    if (method != "rsassa-pss" && method != "rsassa-pss-sha256" && method != "ed25519") {
+      throw SecurityException(repository, std::string("Unsupported sign method: ") + (*sig)["method"].asString());
+    }
+    std::string keyid = (*sig)["keyid"].asString();
+    if (!keys_.count(keyid)) {
+      LOGGER_LOG(LVL_info, "Signed by unknown keyid, skipping");
+      LOGGER_LOG(LVL_trace, "Key ID: " << keyid);
+      continue;
+    }
+
+    if (!keys_for_role_.count(std::make_pair(role, keyid))) {
+      LOGGER_LOG(LVL_warning, "KeyId is not valid to sign for this role");
+      LOGGER_LOG(LVL_trace, "KeyId: " << keyid);
+    }
+
+    std::string signature = (*sig)["sig"].asString();
+    if (Crypto::VerifySignature(keys_[keyid], signature, canonical)) {
+      valid_signatures++;
+    } else {
+      LOGGER_LOG(LVL_warning, "Signature was present but invalid");
+      LOGGER_LOG(LVL_trace, "  keyid:" << keyid);
+      LOGGER_LOG(LVL_trace, "  signature:" << signature);
+      // throw SecurityException(repository, "Invalid signature, verification failed");
```

Comment:
I'm unsure what the right thing to do here is.  The security check is based on getting a 'threshold' number of valid signatures, so there isn't a problem if there are signatures that are present but that we don't trust.

Currently we just ignore invalid signatures (i.e. we treat them as if they weren't present, or we didn't have the matching public key).